src/main/scala/loamstream/loam/LoamGraphValidation.scala

```diff
@@ -65,47 +66,47 @@ object LoamGraphValidation {
         newBulkIssue[(Store, Tool), Unit](graph, this, producerEntry, (), Severity.Warning,
           s"Tool $tool is producer of store $store, but the store is not among its outputs."))
     }
   }
 
-  val eachStoresIsInputOfItsConsumers: LoamConsumerRule[Unit] = new LoamConsumerRule[Unit] {
-    override def apply(graph: LoamGraph, consumerEntry: (Store, Tool)):
-    Seq[LoamConsumerIssue[Unit]] = {
+  object eachStoresIsInputOfItsConsumers extends LoamConsumerRule[Unit] {
+    override def apply(graph: LoamGraph, consumerEntry: (Store, Tool)): Seq[LoamConsumerIssue[Unit]] = {
       val (store, tool) = consumerEntry
       issueIf(!graph.toolInputs.getOrElse(tool, Set.empty).contains(store),
         newBulkIssue[(Store, Tool), Unit](graph, this, consumerEntry, (), Severity.Warning,
           s"Tool $tool is consumer of store $store, but the store is not among its inputs."))
     }
   }
-
-  val eachToolsInputStoresArePresent: LoamToolRule[Set[Store]] =
-    new LoamToolRule[Set[Store]] {
-      override def apply(graph: LoamGraph, tool: Tool): Seq[LoamToolIssue[Set[Store]]] = {
-        val missingInputs = graph.toolInputs.getOrElse(tool, Set.empty) -- graph.stores
-        issueIf(missingInputs.nonEmpty,
-          newBulkIssue[Tool, Set[Store]](graph, this, tool, missingInputs, Severity.Warning,
-            s"The following inputs of tool $tool are missing from the graph: ${missingInputs.mkString(", ")}"))
-      }
+  
+  object eachToolsInputStoresArePresent extends LoamToolRule[Set[Store]] {
+    override def apply(graph: LoamGraph, tool: Tool): Seq[LoamToolIssue[Set[Store]]] = {
+      //NB: Use fasterSetDiff because `--` had O(n^2) running time :( 
+      val missingInputs = Sets.fasterSetDiff(graph.toolInputs.getOrElse(tool, Set.empty), graph.stores)
+      
+      issueIf(missingInputs.nonEmpty,
+        newBulkIssue[Tool, Set[Store]](graph, this, tool, missingInputs, Severity.Warning,
+          s"The following inputs of tool $tool are missing from the graph: ${missingInputs.mkString(", ")}"))
     }
+  }
 
-  val eachToolsOutputStoresArePresent: LoamToolRule[Set[Store]] =
-    new LoamToolRule[Set[Store]] {
-      override def apply(graph: LoamGraph, tool: Tool): Seq[LoamToolIssue[Set[Store]]] = {
-        val missingOutputs = graph.toolOutputs.getOrElse(tool, Set.empty) -- graph.stores
-        issueIf(missingOutputs.nonEmpty,
-          newBulkIssue[Tool, Set[Store]](graph, this, tool, missingOutputs, Severity.Warning,
-            s"The following outputs of tool $tool are missing from the graph: ${missingOutputs.mkString(", ")}"))
-      }
+  object eachToolsOutputStoresArePresent extends LoamToolRule[Set[Store]] {
+    override def apply(graph: LoamGraph, tool: Tool): Seq[LoamToolIssue[Set[Store]]] = {
+      //NB: Use fasterSetDiff because `--` had O(n^2) running time :(
+      val missingOutputs = Sets.fasterSetDiff(graph.toolOutputs.getOrElse(tool, Set.empty), graph.stores)
```

Comment:
I get nervous about function names including adjectives like `fast`, `slow`, etc. Why would I ever use the "slow" version? If the "fast" version is so much better, why wouldn't we always use it, and then why call it "fast" as opposed to replacing the existing implementation?

So, I guess my questions are: 

* what makes `fasterSetDiff` actually faster?
* is there already something in the standard library that does what `fasterSetDiff` does that wasn't immediately obvious?
* would it make sense to name it something more specific to what it does (and how it's different from `--`)?

Reply:
After seeing that later the function uses the `HashSet` method, I'd likely rename it to something along those lines: `hashSetDiff` or similar. My 2c.

Reply:
> `hashSetDiff`

That sounds good to me.